File: Heap/SkylineProblem.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
#include <unordered_map>

using std::vector;
using std::priority_queue;
using std::unordered_map;

class Solution {
private:
    struct Event {
        int loc;
        int height;
        int type;

        Event(int a, int b, int c) : loc(a), height(b), type(c) {}

        bool operator<(const Event& e) {
            // if entering, first process highest
            // if leaving, first process shortest
            if (loc == e.loc)
                return height * type > e.height * e.type;
            return loc < e.loc;
        }
    };
public:
    vector<vector<int>> getSkyline(vector<vector<int>>& buildings) {
        vector<Event> events;
        priority_queue<int> pq;
        vector<vector<int>> res;
        unordered_map<int, int> mp;

        for (auto& building : buildings) {
            events.push_back(Event(building[0], building[2], 1));
            events.push_back(Event(building[1], building[2], -1));
        }
        std::sort(events.begin(), events.end());

        for (auto& e: events) {
            if (e.type == 1) {
                // entering
                if (pq.empty() || pq.top() < e.height) 
                    res.push_back({e.loc, e.height});
                pq.push(e.height);
            }
            else {
                // leaving
                // remove val from pq
                mp[e.height]++;
                while (!pq.empty() && mp[pq.top()] > 0) {
                    mp[pq.top()]--;
                    pq.pop();
                }
                if (pq.empty())
                    res.push_back({e.loc, 0});
                else if (e.height > pq.top())
                    res.push_back({e.loc, pq.top()});
            }
        }

        return res;
    }
};

int main()
{
    Solution sol;
    // vector<vector<int>> buildings {{0,2,3},{0,2,4}};
    // vector<vector<int>> buildings {{0,2,3},{2,5,4}};
    // vector<vector<int>> buildings {{0,2,3},{2,5,3}};
    vector<vector<int>> buildings {{2,9,10},{3,7,15},{5,12,12},{15,20,10},{19,24,8}};
    vector<vector<int>> res = sol.getSkyline(buildings);

    for (auto& e: res) {
            std::cout << e[0] << ' ' << e[1] << '\n';
        }

    return 0;
}

File: Heap/SkylineProblem.py
from typing import List
from collections import Counter
import heapq

class Solution:
    def getSkyline(self, buildings: List[List[int]]) -> List[List[int]]:
        events = []
        h = []
        res = []
        rm_set = Counter()

        for l, r, height in buildings:
            # entering
            events.append((l, height, 1))
            # leaving
            events.append((r, height, -1))
            
        # if entering, first process highest
        # if leaving, first process shortest
        events.sort(key=lambda x: (x[0], -x[1]*x[2]))
        # print(events)

        for loc, height, typ in events:
            if typ == 1:
                if not h:
                    res.append((loc, height))
                elif -h[0] < height:
                    res.append((loc, height))
                heapq.heappush(h, -height)
            else:
                # heap remove 
                rm_set[-height] += 1
                while h and rm_set[h[0]] > 0:
                    rm_set[h[0]] -= 1
                    heapq.heappop(h)

                if not h:
                    res.append((loc, 0))
                elif height > -h[0]:
                    res.append((loc, -h[0]))
            # print(loc, height, typ, h)
                    
        return res


if __name__ == '__main__':
    # buildings = [[2,9,10],[3,7,15],[5,12,12],[15,20,10],[19,24,8]]
    buildings = [[0,2,3],[2,5,3]]
    # buildings = [[0,2,3],[0,2,4]]
    sol = Solution()
    res = sol.getSkyline(buildings)
    print(res)